File: 18-4sum/4sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[][]}
 */
var fourSum = function(nums, target) {
    nums.sort((a,b)=>a-b)
    let n = nums.length
    let st = new Set();
    
    for(let i=0;i<n-3;i++){
        for(let j=i+1;j<n-2;j++){
            const ntar = target - nums[i] - nums[j];
            let l = j+1
            let r = n-1
            while(l<r){
                if(nums[l]+nums[r]<ntar) l++
                else if(nums[l]+nums[r]>ntar) r--
                else{
                    st.add([nums[i],nums[j],nums[l],nums[r]].toString());
                    l++
                    r--
                }
            }
        }
    }

    let ans = []
    for(const triplet of st){
        ans.push(triplet.split(',').map(Number));
    }
    return ans
};

File: 18-4sum/4sum.cpp
class Solution {
public:
    //Pepcoding
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
        set<vector<int>>st;
        int n = nums.size();

        sort(nums.begin(),nums.end());

        for(int i=0;i<n;i++){
            for(int j=i+1;j<n;j++){
                int k=j+1,l=n-1;
                while(k<l){
                    long long sum = (long long)nums[i]+nums[j]+nums[k]+nums[l];
                    if(sum==target){
                        st.insert({nums[i],nums[j],nums[k],nums[l]});
                        k++;
                        l--;
                    }else if(sum<target){
                        k++;
                    }else{
                        l--;
                    }
                }
            }
        }

        return vector<vector<int>>(begin(st),end(st));
    }
};